File: tests/typesafety/typescript/nullable_chain.ts
// expect: passes

function parseQuantity(text: string): number | null {
  return /^\d+$/.test(text) ? Number(text) : null;
}

function checkStock(quantity: number): number | null {
  return quantity <= 10 ? quantity : null;
}

export function quote(text: string): number | null {
  const quantity = parseQuantity(text);
  if (quantity === null) {
    return null;
  }
  const inStock = checkStock(quantity);
  if (inStock === null) {
    return null;
  }
  return inStock * 250;
}

File: tests/typesafety/python/nullable_chain.py
# expect: passes
# title: quote returns None for three different reasons
def parse_quantity(text: str) -> int | None:
    return int(text) if text.isdigit() else None


def check_stock(quantity: int) -> int | None:
    return quantity if quantity <= 10 else None


def quote(text: str) -> int | None:
    quantity = parse_quantity(text)
    if quantity is None:
        return None
    in_stock = check_stock(quantity)
    if in_stock is None:
        return None
    return in_stock * 250
